### PR TITLE
Handle nanopore fastq headers

### PR DESCRIPTION
**Pull request by henningonsbring**
## Description
Currently cg assumes all data is illumina data. This PR allows pipelines to start via cg when it is nanopore data.

### Added

- Parsing of nanopore fastq header

### Changed

-

### Fixed

-

### How to prepare for test
- [ ] ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] paxa the environment: `paxa`
- [ ] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh no-qc-external`

### How to test
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/meta/workflow/fastq.py [hunk 1]
```diff
@@ -147,22 +147,45 @@ def parse_header(line: str) -> dict:
 
         return fastq_meta
 
+    @staticmethod
+    def parse_nanopore_header(line: str) -> dict:
+        fastq_meta = {"lane": None, "flowcell": None}
```
Comment (Vince-janv) [inline cg/meta/workflow/fastq.py:152]:
```suggestion
        header_dict = my_dict = dict(iter((value_pair.split("=") for value_pair in line.split(" ")[1:])))
        lane = header_dict["ch"]
        flowcell = header_dict["flow_cell_id"]
```
This should be safer than using magic numbers

File: cg/meta/workflow/fastq.py [hunk 2]
```diff
@@ -147,22 +147,45 @@ def parse_header(line: str) -> dict:
...
+        header_metadata: list = line.split(" ")
+        lane = header_metadata[3].split("=")
+        fastq_meta["lane"] = lane[1]
+        flowcell = header_metadata[5].split("=")
+        fastq_meta["flowcell"] = flowcell[1]
+        return fastq_meta
+
     @staticmethod
     def parse_file_data(fastq_path: Path) -> dict:
```
Comment (Vince-janv) [inline cg/meta/workflow/fastq.py:161]: Is there a way to separate this more into the microsalt workflow? I think including nanopore directly into a function we use for every fastq file makes it a bit harder to understand for someone unfamiliar to it? Maybe the microsalt analysis_api can have its own `gather_file_metadata_for_sample` function with the nanopore exception?

File: cg/meta/workflow/fastq.py (hunk 1 above)
Comment (henningonsbring) [reply cg/meta/workflow/fastq.py:152]: But does it work? There are not equal signs on every "section"

Comment (Vince-janv) [reply cg/meta/workflow/fastq.py:152]: > dict(iter((value_pair.split("=") for value_pair in nanopore_header.split(" ")[1:])))
> {'runid': '86134885e7ff0e1b44adfc2142bcac582dc34dd3', 'read': '12876', 'ch': '449', 'start_time': '2022-03-16T11:25:34.467266+00:00', 'flow_cell_id': 'FAR96930', 'protocol_group_id': 'CS24_NewprimerpoolTest_N1-24_O25-48', 'sample_id': 'CS24', 'barcode': 'barcode01', 'barcode_alias': 'barcode01', 'parent_read_id': 'c94862ea-e60d-4770-8870-fca0243e98d8'}
>

File: cg/meta/workflow/fastq.py (hunk 2 above)
Comment (Vince-janv) [reply cg/meta/workflow/fastq.py:161]: I realised that there is a mutanfastqhandler lower down in the same file. I think these changes should be placed there.

Comment (henningonsbring) [reply cg/meta/workflow/fastq.py:160]: I have moved some to the mutantfastqhandler, but code is very re-written now (had to to make it work), however not all code is under mutantfastqhandler (for cases where I wrote a new function not present before). What do you think?

**Comment by henningonsbring** [comment]
As it is written now the nanopore code is as little invading on the illumina code as possible which makes it easy to remove/rewrite the nanopore-related code without too much complexity.

**Comment by henningonsbring** [comment]
Documenting the `sars-cov-2_123789_results.csv` content (main result from analysis of case noblecat):

```
Sample,Selection,Region Code,Ticket,%N_bases,%10X_coverage,QC_pass,Lineage,PangoLEARN_version,VOC,Mutations
22CS999701,Information saknas,01,123789,2.31,99.17,TRUE,AY.45,2022-03-22,No,D614G;H655Y;P681R
22CS999702,Information saknas,01,123789,0.64,99.4,TRUE,AY.43,2022-03-22,No,D614G
22CS999703,Information saknas,01,123789,4.44,97.39,TRUE,AY.5,2022-03-22,No,D614G;P681R
22CS999704,Information saknas,01,123789,16.93,88.85,FALSE,AY.4,2022-03-22,No,L452R;T478K;D614G;P681R
22CS999705,Information saknas,01,123789,99.81,0.0,FALSE,None,2022-03-22,No,-
22CS999706,Information saknas,01,123789,2.41,98.03,TRUE,AY.43,2022-03-22,No,L452R;D614G
22CS999707,Information saknas,01,123789,34.13,68.42,FALSE,None,2022-03-22,No,L452R;T478K;D614G;P681R
22CS999708,Information saknas,01,123789,22.94,82.39,FALSE,None,2022-03-22,No,G142D;L452R;T478K;D614G;P681R
22CS999709,Information saknas,01,123789,8.01,96.27,TRUE,AY.9.2,2022-03-22,No,D614G
22CS999710,Information saknas,01,123789,29.99,77.75,FALSE,None,2022-03-22,No,G142D;D614G;P681R
22CS999711,Information saknas,01,123789,14.32,90.63,TRUE,AY.122,2022-03-22,No,D614G
22CS999712,Information saknas,01,123789,16.98,85.83,FALSE,AY.4,2022-03-22,No,L452R;D614G;P681R
22CS999713,Information saknas,01,123789,8.45,92.39,TRUE,AY.42,2022-03-22,No,D614G;P681R
22CS999714,Information saknas,01,123789,4.19,98.03,TRUE,AY.9.2,2022-03-22,No,D614G
22CS999715,Information saknas,01,123789,5.85,95.81,TRUE,AY.9.2,2022-03-22,No,D614G;P681R
22CS999716,Information saknas,01,123789,3.48,97.97,TRUE,AY.9.2,2022-03-22,No,D614G
22CS999717,Information saknas,01,123789,12.78,94.65,TRUE,AY.9.2,2022-03-22,No,A222V;L452R;T478K;D614G;P681R
22CS999718,Information saknas,01,123789,6.57,94.85,TRUE,AY.9.2,2022-03-22,No,A222V;D614G;P681R
22CS999719,Information saknas,01,123789,3.54,98.94,TRUE,AY.9.2,2022-03-22,No,Q613H;D614G
22CS999720,Information saknas,01,123789,5.87,95.75,TRUE,AY.9.2,2022-03-22,No,G142D;A222V
22CS999721,Information saknas,01,123789,63.28,44.56,FALSE,None,2022-03-22,No,L452R;T478K
22CS999722,Information saknas,01,123789,10.03,93.77,TRUE,AY.9.2,2022-03-22,No,A222V
22CS999723,Information saknas,01,123789,7.98,97.47,TRUE,AY.122,2022-03-22,No,-
22CS999724,Information saknas,01,123789,48.36,51.61,FALSE,None,2022-03-22,No,-
22CS999725,Information saknas,01,123789,2.45,99.12,TRUE,AY.9.2,2022-03-22,No,A222V
22CS999726,Information saknas,01,123789,2.48,99.4,TRUE,AY.9.2,2022-03-22,No,A222V;D614G;P681R
22CS999727,Information saknas,01,123789,4.12,98.6,TRUE,AY.9.2,2022-03-22,No,G142D;A222V
22CS999728,Information saknas,01,123789,10.1,93.0,TRUE,AY.9.2,2022-03-22,No,G142D;A222V
22CS999729,Information saknas,01,123789,4.19,95.81,TRUE,AY.9.2,2022-03-22,No,A222V
22CS999730,Information saknas,01,123789,4.86,95.94,TRUE,AY.4,2022-03-22,No,T95I
22CS999731,Information saknas,01,123789,2.93,98.67,TRUE,AY.122,2022-03-22,No,-
22CS999732,Information saknas,01,123789,4.19,98.81,TRUE,AY.9.2,2022-03-22,No,A222V;D614G;P681R
22CS999733,Information saknas,01,123789,31.14,75.53,FALSE,None,2022-03-22,No,G142D;L452R;D614G;P681R
22CS999734,Information saknas,01,123789,4.19,98.19,TRUE,AY.9.2,2022-03-22,No,A222V;Q613H;D614G
22CS999735,Information saknas,01,123789,6.24,94.73,TRUE,AY.42,2022-03-22,No,-
22CS999736,Information saknas,01,123789,5.28,96.62,TRUE,AY.122,2022-03-22,No,D614G
22CS999737,Information saknas,01,123789,3.64,97.2,TRUE,AY.61,2022-03-22,No,D614G
22CS999738,Information saknas,01,123789,7.24,95.33,TRUE,AY.9.2,2022-03-22,No,A222V;D614G;P681R
22CS999739,Information saknas,01,123789,3.57,98.47,TRUE,AY.9.2,2022-03-22,No,-
22CS999740,Information saknas,01,123789,5.67,95.29,TRUE,AY.43,2022-03-22,No,-
22CS999741,Information saknas,01,123789,11.17,96.02,TRUE,AY.9.2,2022-03-22,No,A222V;D614G;P681R
22CS999742,Information saknas,01,123789,9.15,97.38,TRUE,AY.122,2022-03-22,No,D614G;P681R
22CS999743,Information saknas,01,123789,99.81,1.91,FALSE,None,2022-03-22,No,-
22CS999744,Information saknas,01,123789,11.97,93.91,TRUE,AY.9.2,2022-03-22,No,A222V;D614G;P681R
22CS999745,Information saknas,01,123789,8.4,94.74,TRUE,AY.9,2022-03-22,No,A222V;D614G;P681R
22CS999746,Information saknas,01,123789,19.96,84.07,FALSE,AY.86,2022-03-22,No,L452R;D614G;P681R
22CS999747,Information saknas,01,123789,58.12,49.25,FALSE,None,2022-03-22,No,G142D;A222V;L452R;T478K;D614G;P681R
22CS999748,Information saknas,01,123789,77.83,22.04,FALSE,None,2022-03-22,No,L452R;T478K;D614G;P681R
```

File: cg/meta/workflow/analysis.py [hunk 3]
```diff
@@ -312,6 +312,62 @@ def link_fastq_files_for_sample(
             self.fastq_handler.concatenate(linked_reads_paths[read], concatenated_paths[read])
             self.fastq_handler.remove_files(value)
 
+    def get_metadata_for_nanopore_sample(self, sample_obj: models.Sample) -> List[dict]:
```
Comment (Vince-janv) [inline cg/meta/workflow/analysis.py:315]: Similar to the fastqhandlers, both these functions can move to mutant.py under `MutantAnalysisAPI`.

File: cg/meta/workflow/fastq.py [hunk 4]
```diff
@@ -147,6 +147,15 @@ def parse_header(line: str) -> dict:
 
         return fastq_meta
 
+    @staticmethod
```
Comment (Vince-janv) [inline cg/meta/workflow/fastq.py:150]: I think both of these can be moved to `MutantFastqHandler`

File: cg/meta/workflow/analysis.py (hunk 3 above)
Comment (henningonsbring) [reply cg/meta/workflow/analysis.py]: moved and it works

File: cg/meta/workflow/fastq.py (hunk 4 above)
Comment (henningonsbring) [reply cg/meta/workflow/fastq.py]: done

**Comment by henningonsbring** [comment]
Great suggestions @Vince-janv it still works as expected:

```
ls noblecat/fastq/*
noblecat/fastq/barcode01:
01_SE100_22CS999701_1.fastq.gz

noblecat/fastq/barcode02:
01_SE100_22CS999702_1.fastq.gz

noblecat/fastq/barcode03:
01_SE100_22CS999703_1.fastq.gz

noblecat/fastq/barcode04:
01_SE100_22CS999704_1.fastq.gz
```

**Review by Vince-janv — APPROVED** [review]
👌

**Pull request merged**